src/resources/cubemap.js

```diff
@@ -170,13 +167,41 @@ Object.assign(pc, function () {
                         assetCubeMap._levelsEvents[index] = asset || null;
                     }
 
-                    // check if source is actually changed
-                    if (sources[index] !== cubemap._levels[0][index])
-                        levelsUpdated = true;
+                    var cubemap = assetCubeMap.resource;
+
+                    var levelsUpdated = false;
+                    for (var i = 0; i < 6; ++i) {
+                        if (!sources[i]) {
+                            levelsUpdated = false;
+                            break;
+                        }
+                        if (sources[i]._levels[0] !== cubemap._levels[0][i]) {
+                            levelsUpdated = true;
+                        }
+                    }
 
                     // when all faces checked
-                    if (count === 6 && levelsUpdated) {
-                        cubemap.setSource(sources);
+                    if (levelsUpdated) {
+
+                        // build levels
+                        var levels = [];
+                        for (var mip = 0; mip < sources[0]._levels.length; ++mip) {
+                            levels.push(sources.map(function (s) {
+                                return s._levels[mip];
+                            }));
+                        }
+
+                        // reconstruct cubemap with new data
+                        assetCubeMap.resource = new pc.Texture(self._device, {
+                            name: "cubemap-faces",
+                            cubemap: true,
+                            rgbm: sources[0].rgbm,
+                            width: sources[0].width,
+                            height: sources[0].height,
+                            format: sources[0].format,
+                            levels: levels
+                        });
```

Comment:
Does this overwrite a cubemap already set on `assetCubeMap.resource`? Does that leak anything?

Reply:
That's a good point.

The previous approach used texture.setSource() which internally modifies the texture's _width and _height based on the levels passed in.

This approach can't work for non-img sources (due to there being no width and height but also the pixel format is not necessarily RGBA8, _compressed flag could be different etc etc) and so we must construct the texture from scratch.

On first load the cubemap texture being replaced is a 4x4 blank texture created when cubemap loading starts. Ideally this just wouldn't be created in the first place. However there may be users depending on that behaviour.

I'm not sure that it's safe to call assetCubeMap.resource.destroy on the old cubemap because that texture may be in use elsewhere. Is there another option?

Reply:
On second thought, I think destroying the old texture should be fine. Code updated accordingly.

Reply:
I don't know, sorry. 😟  But I'm not sure why somebody would rely on the 4x4 blank texture behavior...